### DISCUSSION TITLE
`group.each.setup` and `group.each.teardown` not working

### DISCUSSION THREAD

**Discussion by backspacerhino** [adonisjs/core#4609] [Japa] [answered]
**Code and setup**

I am trying to delete all from db and migrate after each test in test group to have clean db between tests. I am following this https://docs.adonisjs.com/guides/testing/database#truncate-tables and this is my code(just to showcase the problem)

```
import UserFactory from '#domains/User/v1/Factories/UserFactory'
import User from '#domains/User/v1/Models/User'
import testUtils from '@adonisjs/core/services/test_utils'
import { test } from '@japa/runner'

test.group('Test', (group) => {

  const userFactory = new UserFactory()

  group.each.setup(() => {
    testUtils.db().migrate()
  })

  test('test 1', async ({}) => {
    const userMock = userFactory.create()
    await User.create(userMock)
    const all = await User.all()
    console.log(all.length)
  })
})
```

**Problem**

I am expecting `console.log(all.length)` to always be `1` because I am expecting  ` testUtils.db().migrate()` to make my db clean between tests. However, my DB does not get cleaned between tests and I get increasing numbers on each test run (1,2,3,4,5,6...). If I put `console.log` in `setup` and `teardown` they are being logged.

I have tried a few different ways:

```
  group.each.setup(() => {
    testUtils.db().migrate()
  })
```
When running `node ace test` it does not exit the test execution and just hangs. After manually exiting the process, DB is not cleaned

```
  group.each.setup(async () => {
    await testUtils.db().migrate()
  })
```
When running `node ace test` it does exit the test execution but DB is not cleaned

```
  group.each.setup(() => {
    testUtils.db().truncate()
  })
```
When running `node ace test` it does not exit the test execution and just hangs. After manually exiting the process, DB is not cleaned

```
  group.each.setup(async () => {
    await testUtils.db().truncate()
  })
```
When running `node ace test` it does exit the test execution but DB is not cleaned

```
  group.each.teardown(() => {
    testUtils.db().migrate()
  })
```
When running `node ace test` it does not exit the test execution and just hangs. After manually exiting the process, DB is not cleaned

```
  group.each.teardown(async () => {
    await testUtils.db().migrate()
  })
```
When running `node ace test` it does exit the test execution but DB is not cleaned

```
  group.each.teardown(() => {
    testUtils.db().truncate()
  })
```
When running `node ace test` it does not exit the test execution and just hangs. After manually exiting the process, DB is not cleaned

```
  group.each.teardown(async () => {
    await testUtils.db().truncate()
  })
```
When running `node ace test` it does exit the test execution but DB is not cleaned

I've also tried with both `setup` and `teardown` and all possible other combinations (both w/o async, both w/o migrate and truncate....)

I haven`t touched `bootstrap.ts`

package.json

```
  "devDependencies": {
    "@adonisjs/assembler": "^7.6.1",
    "@adonisjs/eslint-config": "^1.3.0",
    "@adonisjs/prettier-config": "^1.3.0",
    "@adonisjs/tsconfig": "^1.3.0",
    "@japa/api-client": "^2.0.3",
    "@japa/assert": "^3.0.0",
    "@japa/plugin-adonisjs": "^3.0.1",
    "@japa/runner": "^3.1.4",
    "@swc/core": "^1.5.7",
    "@types/luxon": "^3.4.2",
    "@types/node": "^20.12.12",
    "eslint": "^8.57.0",
    "hot-hook": "^0.2.6",
    "pino-pretty": "^11.1.0",
    "prettier": "^3.2.5",
    "ts-node": "^10.9.2",
    "typescript": "^5.4.5"
  },
  "dependencies": {
    "@adonisjs/auth": "^9.2.1",
    "@adonisjs/core": "^6.9.0",
    "@adonisjs/cors": "^2.2.1",
    "@adonisjs/lucid": "^20.6.0",
    "@vinejs/vine": "^2.0.0",
    "argon2": "^0.40.3",
    "luxon": "^3.4.4",
    "mysql2": "^3.10.0",
    "reflect-metadata": "^0.2.2"
  },
```

**Comment by RomainLanz** [comment]
Hey @backspacerhino! 👋🏻 

The code `testUtils.db().migrate()` will migrate the database, and once it is migrated, it will output "Nothing to be change".

What you are looking for is global transaction or truncating the table between each test.

```ts
test.group('User', (group) => {
  group.each.setup(() => testUtils.db().withGlobalTransaction())
})
```

```ts
test.group('User', (group) => {
  group.each.setup(() => testUtils.db().truncate())
})
```

**Reply by backspacerhino** [reply]
Hey @RomainLanz 

I've tried with `withGlobalTransaction` and  `truncate` (as described in the original post) but it is not working, when using `withGlobalTransaction` with or without `async` process hangs but after manually exiting the `length` stays the same(expected). When using `truncate` it behaves as described in original post

**Comment by backspacerhino** [comment, marked as answer]
I've found the solution.

```
  group.each.setup(async () => {
    const func = await testUtils.db().truncate()
    await func()
  })
```

If you do `console.log(func)` you will get `[Function (anonymous)]`. And if you `console.log(func())` you will get `Promise { <pending> }` so it makes sense to `await` but it also works with only `func()`.

However it does not work without `async` because `func` is a promise.

@RomainLanz  this does not seem like how it should work. If you point me to the right direction I can check the source code for this and check why it is behaving like this and/or change the docs.

**Reply by eduwass** [reply]
Ran into this issue as well, the double await was in fact the solution.

Would be nice to update the [docs here](https://docs.adonisjs.com/guides/testing/database#truncate-tables-after-each-run-cycle) @RomainLanz

**Reply by eduwass** [reply]
Quick PR to update docs:
https://github.com/adonisjs/v6-docs/pull/172

**Comment by lgriffith10** [comment]
Hello I've got the same issue right now
The selected issue isn't working on my side, any updates about this ? 

Thanks
